internal/pkg/build/metadata/metadata.go

```diff
@@ -0,0 +1,169 @@
+
+	if fimg != nil {
+		var err error
+		// Get the primary partition data size
+		primSize := make([]*sif.Descriptor, 1)
+		primSize[0], _, err = fimg.GetPartPrimSys()
+		if err != nil {
+			return fmt.Errorf("failed getting main data: %s", err)
+		}
+		labels["system-partition"]["org.label-schema.image-size"] = readBytes(float64(primSize[0].Storelen))
+
+		// Get the image arch
+		imgParts, _, err := fimg.GetPartFromGroup(sif.DescrDefaultGroup)
+		if err != nil {
+			return fmt.Errorf("unable to get image part: %s", err)
+		}
+
+		if len(imgParts) != 1 {
+			sylog.Warningf("Multiple partitions found, using first")
+		}
+
+		imageArch, err := imgParts[0].GetArch()
+		if err != nil {
+			return fmt.Errorf("unable to get image arch: %s", err)
+		}
+		labels["system-partition"]["org.label-schema.image-arch"] = sif.GetGoArch(cstrToString(imageArch[:]))
+	}
+
+	if b != nil {
+		// help info if help exists in the definition and is run in the build
+		if b.RunSection("help") && b.Recipe.ImageData.Help.Script != "" {
+			labels["system-partition"]["org.label-schema.usage"] = "/.singularity.d/runscript.help"
+			labels["system-partition"]["org.label-schema.usage.singularity.runscript.help"] = "/.singularity.d/runscript.help"
+		}
+
+		// bootstrap header info, only if this build actually bootstrapped
+		if !b.Opts.Update || b.Opts.Force {
+			for key, value := range b.Recipe.Header {
+				labels["system-partition"]["org.label-schema.usage.singularity.deffile."+key] = value
+			}
+		}
+	}
+
+	return nil
+}
+
+// copy-paste from sylabs/sif
+func cstrToString(str []byte) string {
+	n := len(str)
+	if m := n - 1; str[m] == 0 {
+		n = m
+	}
+	return string(str[:n])
+}
+
+// TODO: put in a common package
+func readBytes(in float64) string {
+	i := 0
+	size := in
+
+	units := []string{"B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"}
+
+	for size > 1024 {
+		size /= 1024
+		i++
+	}
+	buf := fmt.Sprintf("%.*f %s", i, size, units[i])
+
+	return buf
+}
+
+// AddLabelPartition will add a label partition to a SIF image.
+func AddLabelPartition(fimg *sif.FileImage, link uint32, data []byte) error {
+	descr, err := getDescr(fimg)
+	if err != nil {
+		return fmt.Errorf("no primary partition found: %s", err)
+	}
+
+	labelPart := sif.DescriptorInput{
+		Datatype: sif.DataLabels,
+		Groupid:  descr[0].Groupid,
+		Link:     link,
+		Fname:    "image-metadata",
+		Data:     data,
+	}
+	labelPart.Size = int64(binary.Size(labelPart.Data))
+
+	err = fimg.AddObject(labelPart)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getDescr(fimg *sif.FileImage) ([]*sif.Descriptor, error) {
+	descr := make([]*sif.Descriptor, 1)
+	var err error
+
+	descr[0], _, err = fimg.GetPartPrimSys()
+	if err != nil {
+		return nil, fmt.Errorf("no primary partition found")
+	}
+
+	return descr, nil
+}
+
+// GetSIFData will return a dataType from a SIF.
+func GetSIFData(fimg *sif.FileImage, dataType sif.Datatype) ([]*sif.Descriptor, error) {
```

Comment:
Doesn't look like it's doing that. It looks like its returning _all_ descriptors of a SIF that are of a particular `dataType`. If this is intended to be a nice util function, maybe have one specific for labels, and one specific for defintions, otherwise this is mostly just a wrapping of `GetLinkedDescrsByType()` and doesn't seem that useful. You can make a nice helper to check for a system partition since that's the only extra thing this is doing, but I have a feeling may that exist somewhere already.

Reply:
Okay, I'll just remove this helper function. Since I'v cleaned the code, and removed extra code, this function got a lot smaller, its not needed anymore.